### PR TITLE
feat: runs e2e test on 3 aks versions

### PR DESCRIPTION
**Pull request by nilekhc**
**Reason for Change**:

Runs e2e test on 3 aks versions in the order of `latest-2` -> `latest-1` -> `latest`

**Requirements**

- [x] squashed commits
- [ ] included documentation
- [ ] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

fixes https://github.com/Azure/secrets-store-csi-driver-provider-azure/issues/480

**Does this change contain code from or inspired by another project?**

- [ ] Yes
- [x] No

**If "Yes," did you notify that project's maintainers and provide attribution?**

**Special Notes for Reviewers**:

### CODE REVIEW

File: test/aks-version.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+func main() {
+	flag.Parse()
+
+	aksVersions := []string{}
+	currentVersion := ""
+
+	aksProfiles := AksProfiles{}
+	json.Unmarshal([]byte(*upgradeProfile), &aksProfiles)
```
Comment (aramase) [inline test/aks-version.go:34]: We should handle errors and fail if something goes wrong

**Review by aramase — COMMENTED** [review]
Could we just use a combination of `az aks` commands and jq to get the 3 possible upgrade versions?

1. To get the first available k8s version
```bash
➜ az aks get-versions -l southcentralus --query "[orchestrators[0].orchestratorVersion]" -o json | jq '.[0]'
WARNING: The behavior of this command has been altered by the following extension: aks-preview
"1.18.14"
```
2. To get the next 2 upgrade versions
```bash
➜ az aks get-versions -l southcentralus --query "[orchestrators[].upgrades[-1].orchestratorVersion]" -o json | jq '.[] | unique'
WARNING: The behavior of this command has been altered by the following extension: aks-preview
[
  "1.19.9",
  "1.20.5"
]
```

File: .pipelines/e2e-tests-template.yaml [hunk 2]
```diff
@@ -93,6 +95,31 @@ jobs:
 
         - ${{ if and(parameters.testClusterUpgrade, eq(clusterType, 'aks')) }}:
           - template: templates/aks-upgrade.yaml
+            parameters:
+              upgradeVersion: "$(AKS_UPGRADE_VERSION_1)"
+
+          - script: |
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .pipelines/e2e-tests-template.yaml:101]: Good point

[the remark this answers is not in the document]
Comment (nilekhc) [reply .pipelines/e2e-tests-template.yaml]: Done

File: .pipelines/templates/aks-setup.yaml [hunk 3]
```diff
@@ -1,16 +1,48 @@
...
+      echo "##vso[task.setvariable variable=AKS_INSTALL_VERSION]$aksVersion"
+    displayName: "Determine aks install versions"
+    condition: and(succeeded(), eq(variables['CLUSTER_TYPE'], 'aks'))
+
+  - ${{ if eq(parameters.testClusterUpgrade, true) }}: #Overrride AKS_INSTALL_VERSION if testing with k8s upgrade
+    - script: |
+        aksProfile=$(az aks get-versions -l $(AZURE_LOCATION) --query 'reverse(sort_by(orchestrators, &orchestratorVersion)[*].{version: orchestratorVersion, upgrades: upgrades})' | jq -c)
+        echo "AKS Profile - $aksProfile"
+  
+        aksVersions=$(go run $(System.DefaultWorkingDirectory)/test/aks-version.go -upgrade-profile="$aksProfile")
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .pipelines/templates/aks-setup.yaml]: file removed

File: test/aks-version.go (hunk 1 above)
Comment (nilekhc) [reply test/aks-version.go:34]: file removed

**Comment by nilekhc** [comment]
> Could we just use a combination of `az aks` commands and jq to get the 3 possible upgrade versions?
> 
> 1. To get the first available k8s version
> 
> ```shell
> ➜ az aks get-versions -l southcentralus --query "[orchestrators[0].orchestratorVersion]" -o json | jq '.[0]'
> WARNING: The behavior of this command has been altered by the following extension: aks-preview
> "1.18.14"
> ```
> 
> 1. To get the next 2 upgrade versions
> 
> ```shell
> ➜ az aks get-versions -l southcentralus --query "[orchestrators[].upgrades[-1].orchestratorVersion]" -o json | jq '.[] | unique'
> WARNING: The behavior of this command has been altered by the following extension: aks-preview
> [
>   "1.19.9",
>   "1.20.5"
> ]
> ```

@aramase I had same thought initially. After revisiting this, I changed implementation to work with `az aks get-upgrades` instead of `az aks get-versions`. With this we are now leveraging `jmespath` queries and `jq` and do not need custom Go code anymore.

File: .pipelines/templates/e2e-test.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      SECRET_NAME: $(SECRET_NAME)
+      SUBSCRIPTION_ID: $(SUBSCRIPTION_ID)
+      TENANT_ID: $(TENANT_ID)
+      IS_UPGRADE_TEST: ${{ parameters.testClusterUpgrade }}
+      ${{ if parameters.resetImageVersion }}:
+        CONFIG_IMAGE_VERSION: ""
+      ${{ if parameters.isBackwardCompatibilityTest }}:
+        IS_BACKWARD_COMPATIBILITY_TEST: ${{ parameters.isBackwardCompatibilityTest }}
+      ${{ if eq(parameters.osType, 'windows') }}:
+        TEST_WINDOWS: true
```
Comment (aramase) [inline .pipelines/templates/e2e-test.yaml:53]: nit: new line

Comment (nilekhc) [reply .pipelines/templates/e2e-test.yaml]: fixed

File: .pipelines/templates/aks-upgrade.yaml [hunk 5]
```diff
@@ -1,12 +1,14 @@
 steps:
   - script: |
-      AKS_UPGRADE_VERSION=$(az aks get-upgrades -g ${AZURE_CLUSTER_NAME} -n ${AZURE_CLUSTER_NAME} --query 'max(controlPlaneProfile.upgrades[].kubernetesVersion)' -otsv)
-      echo "##vso[task.setvariable variable=AKS_UPGRADE_VERSION]${AKS_UPGRADE_VERSION}"
-    displayName: "Get upgrade version"
+      #Get minor version of currently installed aks version. major.minor.path -> major.minor
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .pipelines/templates/aks-upgrade.yaml]: fixed

File: .pipelines/templates/aks-setup.yaml [hunk 6]
```diff
@@ -1,16 +1,41 @@
...
+    condition: and(succeeded(), eq(variables['CLUSTER_TYPE'], 'aks'))
+
+  - ${{ if eq(parameters.testClusterUpgrade, true) }}: #Overrride AKS_INSTALL_VERSION if testing with k8s upgrade
+    #If we are running test with cluster upgrade, start with minimum possible non preview version. 
+    - script: |
+        aksVersion=$(az aks get-versions -l westus2 --query "min(orchestrators[?isPreview==null].orchestratorVersion)" -otsv)
+        echo "AKS Install version - $aksVersion"
+
+        echo "##vso[task.setvariable variable=AKS_INSTALL_VERSION]$aksVersion"
+      displayName: "Determine aks install versions for upgrade test"
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .pipelines/templates/aks-setup.yaml]: fixed

File: .pipelines/templates/aks-setup.yaml [hunk 7]
```diff
@@ -1,16 +1,41 @@
...
+    #Run test with latest on preview aks version available
+  - script: | 
+      aksVersion=$(az aks get-versions -l $(AZURE_LOCATION) --query "max(orchestrators[?isPreview==null].orchestratorVersion)" -otsv)
+      echo "AKS Install version - $aksVersion"
+
+      echo "##vso[task.setvariable variable=AKS_INSTALL_VERSION]$aksVersion"
+    displayName: "Determine aks install versions"
+    condition: and(succeeded(), eq(variables['CLUSTER_TYPE'], 'aks'))
+
+  - ${{ if eq(parameters.testClusterUpgrade, true) }}: #Overrride AKS_INSTALL_VERSION if testing with k8s upgrade
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .pipelines/templates/aks-setup.yaml]: fixed

File: .pipelines/templates/aks-setup.yaml [hunk 8]
```diff
@@ -1,16 +1,41 @@
...
+    default: false
+
 steps:
   - script: |
       echo "##vso[task.setvariable variable=AZURE_CLUSTER_NAME]sscd-e2e-$(openssl rand -hex 6)"
       echo ${AZURE_CLUSTER_NAME}
     displayName: "Set cluster name"
     condition: and(succeeded(), eq(variables['CLUSTER_TYPE'], 'aks'))
+    
+    #Run test with latest on preview aks version available
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .pipelines/templates/aks-setup.yaml]: fixed

File: .pipelines/templates/aks-setup.yaml [hunk 9]
```diff
@@ -1,16 +1,41 @@
...
+  - script: | 
+      aksVersion=$(az aks get-versions -l $(AZURE_LOCATION) --query "max(orchestrators[?isPreview==null].orchestratorVersion)" -otsv)
+      echo "AKS Install version - $aksVersion"
+
+      echo "##vso[task.setvariable variable=AKS_INSTALL_VERSION]$aksVersion"
+    displayName: "Determine aks install versions"
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply .pipelines/templates/aks-setup.yaml]: fixed

File: .pipelines/templates/aks-setup.yaml [hunk 10]
```diff
@@ -1,16 +1,42 @@
...
+
+      echo "##vso[task.setvariable variable=AKS_INSTALL_VERSION]$aksVersion"
+    displayName: "Determine aks install version"
+    condition: and(succeeded(), eq(variables['CLUSTER_TYPE'], 'aks'))
+
+  - ${{ if eq(parameters.testClusterUpgrade, true) }}:
+    #Overrride AKS_INSTALL_VERSION if testing with k8s upgrade
+    #If we are running test with cluster upgrade, start with minimum possible non preview version. 
+    - script: |
+        aksVersion=$(az aks get-versions -l westus2 --query "min(orchestrators[?isPreview==null].orchestratorVersion)" -otsv)
```
Comment (aramase) [inline .pipelines/templates/aks-setup.yaml:26]:
```suggestion
        aksVersion=$(az aks get-versions -l $(AZURE_LOCATION) --query "min(orchestrators[?isPreview==null].orchestratorVersion)" -otsv)
```

Comment (nilekhc) [reply .pipelines/templates/aks-setup.yaml]: fixed

**Review by aramase — APPROVED** [review]

**Pull request merged**